File: evd_ros_backend/evd_ros_core/src/evd_script/cache.py
'''
Provides access to any node within EvD by UUID. This is a flattend structure
that is kept coherent with EvDScript's AST.

All nodes can be accessed without a type hint. Additionally, major types can
be more quickly accessed by supplying their hint.
'''


# Cache is global so that we can keep a UUID list for NodeParser
cacheObj = None
def get_evd_cache_obj():
    global cacheObj

    if cacheObj == None:
        cacheObj = Cache()

    return cacheObj


from .data_nodes.trajectory import Trajectory
from .data_nodes.location import Location
from .data_nodes.waypoint import Waypoint
from .data_nodes.thing import Thing
from .data_nodes.trace import Trace
from .data_nodes.machine import Machine
from .data_nodes.thing_type import ThingType
from .data_nodes.regions.region import Region
from .program_nodes.skill import Skill
from .program_nodes.primitive import Primitive
from .environment_nodes.environment_node import EnvironmentNode
from .program import Program
from .context import Context
from .environment import Environment



class Cache(object):

    def __init__(self):
        self.data = {}
        self.instance_table = {
            'trajectory': {'class': Trajectory, 'table': {}},
            'location': {'class': Location, 'table': {}},
            'waypoint': {'class': Waypoint, 'table': {}},
            'thing': {'class': Thing, 'table': {}},
            'trace': {'class': Trace, 'table': {}},
            'machine': {'class': Machine, 'table': {}},
            'environment': {'class': Environment, 'table': {}},
            'program': {'class': Program, 'table': {}},
            'primitive': {'class': Primitive, 'table': {}},
            'context': {'class': Context, 'table': {}},
            'thing_type': {'class': ThingType, 'table': {}},
            'environment_node': {'class': EnvironmentNode, 'table': {}},
            'skill': {'class': Skill, 'table': {}},
            'region': {'class': Region, 'table': {}}
        }

    def add(self, uuid, node):
        self.data[uuid] = node

        for entry in self.instance_table.values():
            if isinstance(node,entry['class']):
                entry['table'][uuid] = node

    def remove(self, uuid):

        node = self.data.pop(uuid, None)

        for entry in self.instance_table.values():
            if isinstance(node,entry['class']):
                entry['table'].pop(uuid, None)

    def clear(self):
        self.data = {}

        for entry in self.instance_table.values():
            entry['table'] = {}

    def get(self, uuid, hint=None):
        retVal = None

        if hint in self.instance_table.keys() and uuid in self.instance_table[hint]['table'].keys():
            retVal = self.instance_table['hint']['table'][uuid]
        else:
            retVal = self.data[uuid]

        return retVal

    def set(self, uuid, dct, hint=None):
        self.get(uuid,hint).set(dct)

    def get_uuids(self, hint=None):
        retVal = None

        if hint in self.instance_table.keys():
            retVal = self.instance_table[hint]['table'].keys()
        elif hint == None:
            retVal = self.data.keys()

        return retVal

    def utility_cache_stats(self):
        log = {'data': {}}

        for n in self.data.values():
            if type(n) not in log['data'].keys():
                log['data'][type(n)] = 0

            log['data'][type(n)] += 1

        for key, entry in self.instance_table.items():
            log['num_' + key] = len(entry['table'])

        return log

File: evd_authoring_app/src/model/evdScript/cache.js

import { Trajectory } from './data/trajectory';
import { Location } from './data/Location';
import { Waypoint } from './data/waypoint';
import { Thing } from './data/thing';
import { Trace } from './data/trace';
import { Machine } from './data/machine';

import { Environment } from './environment/environment';
import { Program } from './program/program';
import { Primitive } from './program/primitive'; 
import { Context } from './context';


// Cache is global so that we can keep a UUID list for NodeParser
cacheObj = null;
export const getEvdCacheObject = () => {
    if (cacheObj == null || cacheObj == undefined) {
        cacheObj = Cache();
    }
    return cacheObj;
}

export class Cache {

    constructor() {
        this.data = {};
        this.trajectories = {};
        this.locations = {};
        this.waypoints = {};
        this.things = {};
        this.traces = {};
        this.machines = {};
        this.environments = {};
        this.programs = {};
        this.primitives = {};
        this.contexts = {};
    }

    add(uuid, node) {
        this.data[uuid] = node;

        if (node instanceof Trajectory) {
            this.trajectories[uuid] = node;
        }

        if (node instanceof Location) {
            this.locations[uuid] = node;
        }

        if (node instanceof Waypoint) {
            this.waypoints[uuid] = node;
        }

        if (node instanceof Thing) {
            this.things[uuid] = node;
        }

        if (node instanceof Trace) {
            this.traces[uuid] = node;
        }

        if (node instanceof Machine) {
            this.machines[uuid] = node;
        }

        if (node instanceof Program) {
            this.programs[uuid] = node;
        }

        if (node instanceof Environment) {
            this.environments[uuid] = node;
        }

        if (node instanceof Primitive) {
            this.primitives[uuid] = node;
        }

        if (node instanceof Context) {
            this.contexts[uuid] = node;
        }
    }

    remove(uuid) {
        node = this.data[uuid];
        delete this.data[uuid];

        if (node instanceof Trajectory) {
            delete this.trajectories[uuid];
        }

        if (node instanceof Location) {
            delete this.locations[uuid];
        }

        if (node instanceof Waypoint) {
            delete this.waypoints[uuid];
        }

        if (node instanceof Thing) {
            delete this.things[uuid];
        }

        if (node instanceof Trace) {
            delete this.traces[uuid];
        }

        if (node instanceof Machine) {
            delete this.machines[uuid];
        }

        if (node instanceof Program) {
            delete this.program[uuid];
        }

        if (node instanceof Environment) {
            delete this.environments[uuid];
        }

        if (node instanceof Primitive) {
            delete this.primitives[uuid];
        }

        if (node instanceof Context) {
            delete this.contexts[uuid];
        }
    }

    clear() {
        this.data = {};
        this.locations = {};
        this.waypoints = {};
        this.trajectories = {};
        this.things = {};
        this.traces = {};
        this.machines = {};
        this.programs = {};
        this.environments = {};
        this.primitives = {};
        this.contexts = {};
    }

    get(uuid, hint=null) {

        let retVal = undefined;

        if (hint === 'trajectory' && uuid in this.trajectories) {
            retVal = this.trajectories[uuid];
        } else if (hint === 'location' && uuid in this.locations) {
            retVal = this.locations[uuid];
        } else if (hint === 'waypoint' && uuid in this.waypoints) {
            retVal = this.waypoints[uuid];
        } else if (hint === 'thing' && uuid in this.things) {
            retVal = this.things[uuid];
        } else if (hint === 'trace' && uuid in this.traces) {
            retVal = this.traces[uuid];
        } else if (hint === 'machine' && uuid in this.machines) {
            retVal = this.machines[uuid];
        } else if (hint === 'program' && uuid in this.programs) {
            retVal = this.programs[uuid];
        } else if (hint === 'environment' && uuid in this.environments) {
            retVal = this.environments[uuid];
        } else if (hint === 'primitive' && uuid in this.primitives) {
            retVal = this.primitives[uuid];
        } else if (hint === 'context' && uuid in this.contexts) {
            retVal = this.contexts[uuid];
        } else {
            retVal = this.data[uuid];
        }

        return retVal;
    }

    set(uuid, dct, hint=null) {
        this.get(uuid,hint).set(dct);
    }

    utilityCacheStats() {
        log = {
            data: {},
            num_trajectories: 0,
            num_locations: 0,
            num_waypoints: 0,
            num_things: 0,
            num_traces: 0,
            num_machines: 0,
            num_primitives: 0,
            num_contexts: 0,
        }

        Object.values(this.data).forEach((node) => {
            if (! node.constructor.name in log.data) {
                log.data[node.constructor.name] = 0;
            }
            log.data[node.constructor.name] += 1;
        });
         
        log.num_trajectories = Object.keys(this.trajectories).length;
        log.num_locations = Object.keys(this.locations).length;
        log.num_waypoints = Object.keys(this.waypoints).length;
        log.num_things = Object.keys(this.things).length;
        log.num_traces = Object.keys(this.traces).length;
        log.num_machines = Object.keys(this.machines).length;
        log.num_primitives = Object.keys(this.primitives).length;
        log.num_contexts = Object.keys(this.contexts).length;
    }

}

